gc/structs/ForwardedHeader.hpp

```diff
@@ -40,7 +42,14 @@
    This hint is not necessary for correctness of copying protocol, it's just an optimization to avoid visiting destination object header
    in cases when it's likely not in data cash (GC thread encountering already forwarded object) */
 #define OMR_BEING_COPIED_HINT 2
+/* Define OMR_SELF_FORWARDED_TAG */
+#if defined(J9_GC_MULTI_SLOT_HOLE)
```

Comment:
hm, why is this needed? 
despite its prefix, J9_GC_MULTI_SLOT_HOLE is actually an OMR const, and should be defined always

Reply:
That's what I thought as well. I've ran several tests and it looks like in some cases `J9_GC_MULTI_SLOT_HOLE` does not get defined (not sure why), and if I don't have that I get something like:
```
    [javac] Compiling 3793 source files to /tmp/bld_464476/job_668560323/debugtools/DDR_VM/output/bin/j9_core
    [javac] /tmp/bld_464476/job_668560323/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCScavengerForwardedHeader_V2.java:39: error: cannot find symbol
    [javac] 	private static final int ALL_TAGS = (int)MM_ForwardedHeader.OMR_FORWARDED_TAG_MASK;
    [javac] 	                                                           ^
    [javac]   symbol:   variable OMR_FORWARDED_TAG_MASK
    [javac]   location: class MM_ForwardedHeader
```

Reply:
Another way to avoid such error was to have something like:
```
#define OMR_SELF_FORWARDED_TAG 1
```
instead of the following and the `#if defs`:
```
#define OMR_SELF_FORWARDED_TAG J9_GC_MULTI_SLOT_HOLE
```
But this would change the logic

Reply:
There is non-obvious connection between J9_GC_MULTI_SLOT_HOLE and OMR_SELF_FORWARDED_TAG...

Forwarded header besides overlapping with class slot of an object, may also overlap with free entry header (in a fragmented heap, when we have a series of interleaved objects and free chunks). J9_GC_MULTI_SLOT_HOLE is used to distinguish between an object (0x1 bit reset) and freechunk (0x1 bit set). This is useful when we want to iterate heap objects in address order....

Since a self-forwarded object will have 0x1 bit set (beside 0x4), iterator  needs a way (the logic is in GC_ObjectHeapIteratorAddressOrderedList::nextObjectNoAdvance()) to distinguish between a free chunk and self-forwarded object. This scenario is rare, but may occur when iterating Evacuate space in a middle of a aborted Scavenge (during abort recovery or when doing a GC check). The space may have all combinations of dead (or more precisely not yet forwarded) objects (0x4 not set, 0x1 not set), strictly forwarded objects (0x4 set, 0x1 not set), self forwarded objects (0x4 set, 0x1 set) and free chunks (0x4 not set and 0x1 set).

So, by doing 
#define OMR_SELF_FORWARDED_TAG J9_GC_MULTI_SLOT_HOLE
we are stating that this bit is overloaded, and may mean both of those things (depending on other flags). Doing just 
#define OMR_SELF_FORWARDED_TAG 1
would be ok, but it would be even less clear that the bit is overloaded

Reply:
I see, how should we define `OMR_SELF_FORWARDED_TAG`? I thought about the `#if #else` approach because that at least keeps `J9_GC_MULTI_SLOT_HOLE` which makes a bit clearer what `OMR_SELF_FORWARDED_TAG` represents